PowerEditor/src/tools/NppShell/Installer.cpp

```diff
@@ -0,0 +1,339 @@
+
+    if (lResult != ERROR_SUCCESS)
+    {
+        return lResult;
+    }
+
+    lResult = RegSetKeyValue(regKey, NULL, name.empty() ? NULL : name.data(), REG_SZ, value.data(), static_cast<DWORD>(value.length() * sizeof TCHAR));
+    
+    RegCloseKey(regKey);
+
+    return lResult;
+}
+
+LRESULT CleanupRegistry(const wstring guid)
+{
+    constexpr int bufferSize = MAX_PATH + GUID_STRING_SIZE;
+    WCHAR buffer[bufferSize];
+    const auto arraySize = bufferSize * sizeof(WCHAR);
+
+    LRESULT result;
+
+    result = RemoveRegistryKeyIfFound(HKEY_CLASSES_ROOT, ShellExtensionKey);
+
+    if (result != ERROR_SUCCESS)
+    {
+        return result;
+    }
+
+    StringCbPrintf(buffer, arraySize, L"Notepad++_file\\shellex");
+    result = RemoveRegistryKeyIfFound(HKEY_CLASSES_ROOT, buffer);
+    if (result != ERROR_SUCCESS)
+    {
+        return result;
+    }
+
+    StringCbPrintf(buffer, arraySize, L"Software\\Classes\\CLSID\\%s", guid.c_str());
+    result = RemoveRegistryKeyIfFound(HKEY_LOCAL_MACHINE, buffer);
+    if (result != ERROR_SUCCESS)
+    {
+        return result;
+    }
+
+    return ERROR_SUCCESS;
+}
+
+LRESULT CleanupHack()
+{
+    wstring keyName = L"SOFTWARE\\Classes\\*\\shell\\pintohome";
+    wstring valueName = L"MUIVerb";
+
+    LRESULT result = 0;
+    bool found = false;
+
+    HKEY hkey;
+    HRESULT hResult = RegOpenKeyEx(HKEY_LOCAL_MACHINE, keyName.c_str(), 0, KEY_READ, &hkey);
+
+    if (hResult == ERROR_SUCCESS)
+    {
+        DWORD cbData = 0;
+        hResult = RegGetValue(hkey, nullptr, valueName.c_str(), RRF_RT_REG_SZ, nullptr, nullptr, &cbData);
+
+        if (hResult == ERROR_SUCCESS)
+        {
+            vector<BYTE> buffer(cbData + 1);
+
+            cbData = (DWORD)buffer.size();
+            hResult = RegGetValue(hkey, nullptr, valueName.c_str(), RRF_RT_REG_SZ, nullptr, &buffer[0], &cbData);
+
+            wstring valueString(reinterpret_cast<wchar_t*>(&buffer[0]));
+
+            found = valueString.find(L"Notepad++") != wstring::npos;
+        }
+    }
+
+    RegCloseKey(hkey);
+
+    if (found)
+    {
+        result = RegDeleteTree(HKEY_LOCAL_MACHINE, keyName.c_str());
+    }
+
+    return result;
+}
+
+HRESULT NppShell::Installer::RegisterSparsePackage()
+{
+    PackageManager packageManager;
+    AddPackageOptions options;
+
+    const wstring externalLocation = GetInstallationPath();
+    const wstring sparsePkgPath = externalLocation + L"\\NppShell.msix";
+
+    Uri externalUri(externalLocation);
+    Uri packageUri(sparsePkgPath);
+
+    options.ExternalLocationUri(externalUri);
+
+    auto deploymentOperation = packageManager.AddPackageByUriAsync(packageUri, options);
+    auto deployResult = deploymentOperation.get();
+
+    if (!SUCCEEDED(deployResult.ExtendedErrorCode()))
+    {
+        return deployResult.ExtendedErrorCode();
+    }
+
+    return S_OK;
+}
+
+HRESULT NppShell::Installer::UnregisterSparsePackage()
+{
+    PackageManager packageManager;
```

Comment:
You got a point, I'll make sure it only tries to unregister it if it is on Win11

Reply:
You might also need to delay load a few libs, as some of the statically linked C APIs for WinRT (`RoGetActivationFactory`) will also be missing. I think cppwinrt has some level of support for building a DLL that can also run on 7, so it might not actually be needed, but better to directly test on a Windows 7 machine/VM.